File: DynamicPrograming/php/validCompound.php
<?php

declare(strict_types=1);

/** @param string[] $elements */
function validCompound(string $compound, array $elements): bool
{
    $lowerCompound = strtolower($compound);
    $lowerElements = array_map('strtolower', $elements);
    $memo = [];

    $dfs = function (int $index) use (&$dfs, &$memo, $lowerCompound, $lowerElements): bool {
        if ($index === strlen($lowerCompound)) {
            return true;
        }
        if (array_key_exists($index, $memo)) {
            return $memo[$index];
        }

        foreach ($lowerElements as $element) {
            $length = strlen($element);
            if (substr($lowerCompound, $index, $length) === $element && $dfs($index + $length)) {
                return $memo[$index] = true;
            }
        }

        return $memo[$index] = false;
    };

    return $dfs(0);
}

File: DynamicPrograming/python/validCompound.py
from functools import lru_cache
from typing import Iterable


def validCompound(compound: str, elements: Iterable[str]) -> bool:
    lower_compound = compound.lower()
    lower_elements = [element.lower() for element in elements]

    @lru_cache(maxsize=None)
    def dfs(i: int) -> bool:
        if i == len(lower_compound):
            return True
        for element in lower_elements:
            if lower_compound.startswith(element, i) and dfs(i + len(element)):
                return True
        return False

    return dfs(0)
